File: banco_problemas_aplicados/nivel_intermedio/soluciones/cpp/problema33.cpp
/*
 * Problema 33 - Compresion de Datos por Submuestreo
 * Toma las muestras en indices 0, K, 2K, ... Devuelve el nuevo tamano.
 */
#include <iostream>
#include <iomanip>
using namespace std;

int submuestrea(double arr[], int n, int k, double res[]) {
    int j = 0;
    for (int i = 0; i < n; i = i + k) {
        res[j] = arr[i];
        j = j + 1;
    }
    return j;
}

int main() {
    int n, k;
    cout << "Ingrese el numero de muestras: ";
    cin >> n;

    double* arr = new double[n];
    for (int i = 0; i < n; i++) {
        cout << "Muestra " << i << ": ";
        cin >> arr[i];
    }

    cout << "Ingrese el factor de compresion K: ";
    cin >> k;

    double* res = new double[n];
    int nuevo = submuestrea(arr, n, k, res);

    cout << fixed << setprecision(2);
    cout << "Vector comprimido (" << nuevo << " muestras): ";
    for (int i = 0; i < nuevo; i++) {
        cout << res[i] << " ";
    }
    cout << endl;

    delete[] arr;
    delete[] res;
    return 0;
}

File: banco_problemas_aplicados/nivel_intermedio/soluciones/java/Problema33.java
/*
 * Problema 33 - Compresion de Datos por Submuestreo
 * Toma las muestras en indices 0, K, 2K, ... Devuelve el nuevo tamano.
 */
import java.util.Scanner;
import java.util.Locale;

public class Problema33 {

    static int submuestrea(double[] arr, int n, int k, double[] res) {
        int j = 0;
        for (int i = 0; i < n; i = i + k) {
            res[j] = arr[i];
            j = j + 1;
        }
        return j;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Ingrese el numero de muestras: ");
        int n = sc.nextInt();

        double[] arr = new double[n];
        for (int i = 0; i < n; i++) {
            System.out.print("Muestra " + i + ": ");
            arr[i] = sc.nextDouble();
        }

        System.out.print("Ingrese el factor de compresion K: ");
        int k = sc.nextInt();

        double[] res = new double[n];
        int nuevo = submuestrea(arr, n, k, res);

        System.out.print("Vector comprimido (" + nuevo + " muestras): ");
        for (int i = 0; i < nuevo; i++) {
            System.out.printf(Locale.US, "%.2f ", res[i]);
        }
        System.out.println();
        sc.close();
    }
}
